Refs #29444 -- Allowed returning multiple fields from INSERT statements on PostgreSQL.

codingjoe opened the pull request:
Ticket [#29444](https://code.djangoproject.com/ticket/29444)

Currently the PK is hardcoded as the only retuning object which is not necessary and currently blocking [#27452](https://code.djangoproject.com/ticket/27452).

### Side note:

**This change should be fully backwards compatible.** I am aware that this fix allows adding database defaults to all backends but MySQL. This is why **I want to keep this an undocumented API** for now.
People will find out about it anyways.

You can implement a database default as follows:

```python
class Default:
    """
    Expression for ``DEFAULT``.

    In an insert query this will return the database default.
    """

    def as_sql(self, compiler, connection):
        return connection.ops.pk_default_value(), []


class MyModel(models.Model):
    my_uuid_with_db_default = models.UUIDField(default=Default)
```

If you have a default set for your column in the database, it will return that instead of trying to set null.
You no longer need to call `refresh_from_db` or any other strange thing.

apollo13 commented on django/db/models/fields/__init__.py:
```diff
@@ -136,12 +136,15 @@ def __init__(self, verbose_name=None, name=None, primary_key=False,
         self.verbose_name = verbose_name  # May be set by set_attributes_from_name
         self._verbose_name = verbose_name  # Store original for deconstruction
         self.primary_key = primary_key
+        self.returning = returning
```
`self.returning = returning or primary_key`

apollo13 commented on django/db/models/options.py:
```diff
@@ -825,3 +825,12 @@ def _property_names(self):
+            field for field in self._get_fields(
+                forward=True, reverse=False, include_parents=PROXY_PARENTS
+            )
+            if getattr(field, 'returning', False)
```
you can drop the `False` to getattr here, `None` is also not true.

codingjoe replied:
Nope, this will raise an AttributeError in Py3 ;)
https://docs.python.org/3/library/functions.html#getattr

apollo13 commented on tests/postgres_tests/test_returning.py:
```diff
@@ -0,0 +1,46 @@
+
+    def test_returning_value_on_obj(self):
+        # ReturningModel uses database functions not python defaults
+        obj = ReturningModel()
```
wouldn't `obj = ReturningModel.objects.create()` save a line?

apollo13 commented on tests/postgres_tests/test_returning.py:
```diff
@@ -0,0 +1,46 @@
+        obj = ReturningModel()
+        obj.save()
+        self.assertIsNotNone(obj.pk)
+        self.assertIsInstance(obj.created, datetime.datetime)
```
That feels weird; I get what you are trying to do here, but could you set the default to a static database value and then check the actual value?

codingjoe replied:
I agree, but it would mean I would have to introduce actual database defaults here. I will keep this once for another PR. If you know a better DB function I can use, let me know.

codingjoe replied:
I could build a workaround, but it would result in more code for setup.

apollo13 commented on tests/postgres_tests/test_returning.py:
```diff
@@ -0,0 +1,46 @@
+        self.assertIsInstance(obj.created, datetime.datetime)
+
+    def test_mti(self):
+        obj = MTIReturning()
```
Same as above

apollo13 commented:
I really like those changes and how small the actual diff is! Keep up the fantastic work

apollo13 commented:
We probably should change `fetch_returned_insert_id` to a name more fitting ala `fetch_returned_data` or so -- or even separate that based on db support, dunno yet.

adamchainz commented on tests/postgres_tests/test_returning.py:
```diff
@@ -0,0 +1,46 @@
+                      ', "postgres_tests_returningmodel"."created"', sql)
+
+    def test_returning_value_on_obj(self):
+        # ReturningModel uses database functions not python defaults
```
idk if this comment has value? the model definition already shows this

codingjoe commented:
> We probably should change fetch_returned_insert_id to a name more fitting ala fetch_returned_data or so -- or even separate that based on db support, dunno yet.

@apollo13 I wanted to keep the changeset small to get the concept across.
Backwards compatibility with 3rd party backends is a good questions. If I am not mistaken this is not public API. That being said. Maybe it makes sense to rename the methods. This ways 3rd party backends can simply implement both methods.

apollo13 commented:
Right, not sure what I was thinking (if at all) 🤣

On May 27, 2018 4:46:56 PM GMT+02:00, Johannes Hoppe <notifications@github.com> wrote:
>codingjoe commented on this pull request.
>
>
>
>> @@ -825,3 +825,12 @@ def _property_names(self):
>             if isinstance(attr, property):
>                 names.append(name)
>         return frozenset(names)
>+
>+    @cached_property
>+    def returning(self):
>+        return [
>+            field for field in self._get_fields(
>+                forward=True, reverse=False,
>include_parents=PROXY_PARENTS
>+            )
>+            if getattr(field, 'returning', False)
>
>Nope, this will raise an AttributeError in Py3 ;)
>https://docs.python.org/3/library/functions.html#getattr
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/9983#discussion_r191078837

timgraham commented:
I don't think we would merge this without documentation, even if MySQL is unsupported. By the way, `postgres_tests` is for testing things in `contrib.postgres`, however, there are no changes in that module so the test location seems incorrect. You can document database backend API changes in the existing section of the release notes.

codingjoe commented:
Ok, so does that mean I can refactor the API a bit?

timgraham commented:
Sure, maybe put it in a separate commit at least for now to ease review.

codingjoe commented:
Ok, documented and renamed. I also moved the tests.

timgraham commented on tests/backends/postgresql/test_returning.py:
```diff
@@ -0,0 +1,44 @@
+        compiler.return_id = True
+        sql = compiler.as_sql()[0][0]
+        self.assertIn('RETURNING "backends_returningmodel"."id"'
+                      ', "backends_returningmodel"."created"', sql)
```
Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

apollo13 commented on django/db/models/sql/compiler.py:
```diff
@@ -1289,25 +1291,26 @@ def as_sql(self):
-                cursor, self.query.get_meta().db_table, self.query.get_meta().pk.column
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            if isinstance(self.query.get_meta().pk, AutoField):
```
Do we need the check for `AutoField` or can we just return `last_insert_id` unconditionally like before?

codingjoe replied:
Yes, we do for MySQL and SQLite in cases where people chose custom primary keys that are not AI (auto increment) integer fields.

codingjoe replied:
Maybe I add a comment ;)

timgraham replied:
Maybe we need a field property. It doesn't look nice to put `isinstance()` special cases here.

codingjoe commented on docs/ref/models/fields.txt:
```diff
@@ -369,6 +369,17 @@ underscores to spaces. See :ref:`Verbose field names <verbose-field-names>`.
+the value is set to the same value as :attr:`~Field.primary_key`.
+
+.. note::
+    Currently only PostgreSQL supports multiple return columns.
```
These are all very valid points. Honestly I kind of don't want this to be a documented feature. It is something I don't recommend using. There are too many implications one needs to be aware of.
I would like to add a more comprehensive documentation when adding the first fields in `django.contrib.postgres` that do use this behavior.

codingjoe replied:
I did remove the documentation for now and would keep the `returning` API undocumented. I will share a bit more light into how to build those things here #7525 
I feel to really understand this, it needs an example.

timgraham commented on django/db/models/options.py:
```diff
@@ -826,3 +826,12 @@ def _property_names(self):
         return frozenset(names)
+
+    @cached_property
+    def returning(self):
```
I guess this should be something like `returning_fields` -- `returning` looks like a boolean attribute as used elsewhere.

codingjoe commented:
Ok, I:
* added documentation for all db backend API change
* removed all non relevant tests
* removed the AutoField import dependency
* rebased and squashed all chages

@timgraham you would mind to give it one more look?

amureki commented on tests/backends/models.py:
```diff
@@ -111,3 +112,7 @@ class Author(models.Model):
+
+
+class ReturningModel(models.Model):
+    created = models.DateTimeField(default=Now, returning=True, editable=False)
```
Looks really cool!
Don't we need to document this argument? Here, for example: https://docs.djangoproject.com/en/2.1/ref/models/fields/#field-options

Or documentation comes somehow separately?

codingjoe replied:
I actually removed the documentation. I don't want this to be part of the public API. At least not for now where there is no use case for it.

timgraham replied:
Is any test coverage lost if these tests are removed? As I said previously, "I think it would be best to remove the PostgreSQL tests that suggest the change is enabling some new behavior (which as you say, shouldn't be used outside of Django itself)." Already we have someone looking at the test and wanting to use the functionality. :-)

Perhaps a comment should be added to ` django/db/models/fields/__init__.py` to mark `returning` as private.

codingjoe replied:
@timgraham if this is about adding more models to the test suite. I can try to remove it in a later PR.

codingjoe commented on django/db/models/fields/related.py:
```diff
@@ -807,6 +807,8 @@ def __init__(self, to, on_delete, related_name=None, related_query_name=None,
         )
         kwargs.setdefault('db_index', True)
+        if parent_link:
+            kwargs['returning'] = True
```
Yes, the `parent_link` is part of the multi table inheritance feature. This is also where this is tested. This is merely retaining the same behavior.

apollo13 commented on django/db/models/sql/compiler.py:
```diff
@@ -1289,25 +1290,27 @@ def as_sql(self):
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            if self.query.get_meta().pk.get_internal_type() in ['AutoField', 'BigAutoField']:
+                # The PK could be a custom field. Only AutoFields are supported in MySQL and SQLite.
```
Something along the lines of:
```
For databases which do not support returning clause we need to ask the database for the id. Generally last_insert_id is only supported for serial/auto_incr columns, hence we guard it for auto field.
# TODO: Maybe add a marker to fields that their value can be returned by last_insert_id instead of a type check.
```

timgraham commented:
Why must `last_insert_id` return a list? Should it be renamed to `last_insert_ids` for some future change? If not, it seems it could continue to return a single value and then `execute_sql()` could wrap the call in a list.

codingjoe commented:
> Why must last_insert_id return a list? Should it be renamed to last_insert_ids for some future change? If not, it seems it could continue to return a single value and then execute_sql() could wrap the call in a list.

Good point @timgraham. I reverted the API change and wrapped the call in `execute_sql` as you suggested.

timgraham commented:
buildbot, test on oracle.

timgraham commented:
There are a number of Oracle test failures.

codingjoe commented:
@timgraham I rebased again, now they are now ;)

felixxm commented:
buildbot, test on oracle.

timgraham commented:
The Oracle tests don't run automatically, you need to use [the trigger phrase](https://code.djangoproject.com/wiki/Jenkins). See [OracleTestSetup](https://code.djangoproject.com/wiki/OracleTestSetup) if you need to set up a test environment.

codingjoe commented:
buildbot, test on oracle

codingjoe commented:
​pull-requests-oracle oracle

codingjoe commented:
@timgraham I has some time to fix it. It works for oracle now as well, at least locally it did. Would you mind to trigger the build again?

felixxm commented:
buildbot, test on oracle.

charettes commented on tests/queries/test_returning.py:
```diff
@@ -0,0 +1,29 @@
+from .models import Author, ReturningModel
+
+
+@unittest.skipUnless(connection.vendor == 'postgresql', 'PostgreSQL tests')
```
Shouldn't this be gated by a feature flag instead since it's supposed to work on Oracle as well?

Isn't this missing actual tests for `ReturningModel` inserts?

codingjoe replied:
HI @charettes no, it does not work on oracle, even tho the code suggest it. It is actually not supported by the db client as far as I understood it :/
I'd love to see this work in the future, for now this remains PG only.

codingjoe replied:
So all the other code is just not NOT break Oracle and I added support for non-int primary keys, as the docs suggest.

charettes replied:
I still think that this test should execute the SQL it generates to assert it's working. Right now it's only introspecting it which doesn't add much value.

To be clear I think that the only thing that should be tested here is that `ReturningModel.objects.create()` has both an `id` and `isinstance(r.created, datetime)` assigned.

charettes commented on tests/model_forms/models.py:
```diff
@@ -356,6 +356,9 @@ def __iter__(self):
 
+    class Meta:
+        ordering = ('name',)
+
```
Is this change related?

codingjoe replied:
I don't remember. Works fine without it now. I'll remove it.

charettes commented on tests/queries/models.py:
```diff
@@ -723,3 +724,9 @@ class Meta:
+class ReturningModel(models.Model):
+    # There's no official support for default=Now or returning=True. This
+    # is merely to test some internal APIs.
+    created = models.DateTimeField(default=Now, returning=True, editable=False)
```
Can you think of a case where you'd want `default=Now(), returning=False`?

What happens if `returning=False`, does the field gets _deferred_ on creation? We certainly don't want `.created` to be a `Now` instance.

What if `returning=True` is passed and the backend doesn't support it? Should we attempt a `SELECT` to retrieve the value or does the field gets deferred.

These are questions that were [discussed in the past](https://github.com/django/deps/pull/36#issuecomment-261151212) but had no clear consensus and have to be tested before we can proceed with this PR.

charettes commented on django/db/models/base.py:
```diff
@@ -883,13 +882,12 @@ def _do_update(self, base_qs, using, pk_val, values, update_fields, forced_updat
         """
-        return manager._insert([self], fields=fields, return_id=update_pk,
-                               using=using, raw=raw)
+        return manager._insert([self], fields=fields, using=using, raw=raw)
```
Is there a reason not to favor the previous approach and pass `returning_fields=self._meta.returning_fields` instead? That seems like a better separation of concerns to me than having the insert compiler lookup `returning_fields`.

codingjoe replied:
What do you mean by previous approach? The argument `rerturn_id` does still exist, but for this call it remains the default.

I honestly don't want to change the API here, I want to keep the changeset as small as possible.

felixxm replied:
I tried to address Simon's request in the 13477049aa6b959a98d6fb941a08edfa13046b31, this change also replaces some misleading variables' names.

felixxm replied:
@charettes Would you like to take a look?

codingjoe replied:
Just one comment. If we go down that path, does that mean, we should also pass the table name in the future? The options are already available to the compiler for that very reason. I don't say it should be done within this patch – in fact I would be against it – but it would make this more consistent and decouple both layers.

felixxm replied:
I don't think that we need to pass a table name because `_insert()` is a `QuerySet` method so it already contains `self.model` with a table name :thinking: but yeah let's leave it for a separate discussion.

charettes commented on django/db/models/fields/related.py:
```diff
@@ -807,6 +807,8 @@ def __init__(self, to, on_delete, related_name=None, related_query_name=None,
         )
         kwargs.setdefault('db_index', True)
+        if parent_link:
+            kwargs['returning'] = True
```
Could be made a property that returns `super().returning or self.remote_field.parent_link`.

codingjoe replied:
I'll move it.

charettes commented on django/db/models/fields/__init__.py:
```diff
@@ -135,12 +135,15 @@ def __init__(self, verbose_name=None, name=None, primary_key=False,
         self.primary_key = primary_key
+        # `returning` is a private API, subject to change. Currently only
+        # PostgreSQL supports returning multiple fields on a model.
+        self.returning = returning or (primary_key and default is NOT_PROVIDED)
```
I'd vote for making `returning` a `property` instead of a stealth field option at least for now because this is not something we've done in the past.

```python
@property
def returning(self):
    return hasattr(self.default, 'as_sql')

AutoField.returning = True
```

That would make `DateTimeField(default=Now)` work and avoid the ambiguity of `default=Now, returning=False`. We'd still have to deal with backends that don't support returning fields.

codingjoe replied:
Sorry, I don't think I fully understand your reasoning. What is the difference between a property or an instance attribute? Isn't the only difference, that the former can't be set?

So what I think you want to achieve, is that you can't set `default=Now` and set returning to False. Right? I think we are getting ahead of ourselves here. This patch does not intent to support database defaults, at least not as a public API.

Anyhow, here is my suggestion. I will remove the `returning` kwarg from the constructor. That way you can't set it at all. It will remain something only exposed to custom fields. Sounds good?

charettes replied:
> So what I think you want to achieve, is that you can't set default=Now and set returning to False. Right?

I'm not saying I want to achieve that. I'm saying that the API you are suggesting, even if private, allows undefined and untested behaviour when `default=Now, returning=False` is set.

> I think we are getting ahead of ourselves here. This patch does not intent to support database defaults, at least not as a public API.

@codingjoe my main concern with adding a model field kwargs is the probability we'll have to deprecate it in the future in favor of something that is more though of and in line with the previous discussions (https://github.com/django/deps/pull/36, #5904, #2149). For example the ORM usually tries to abstract implementation details and prefix database related options with `db_*`.

Adding a stealth model field kwarg or deprecating one is not something we've done in the past and I assume it's going to be a lot more pain than deprecating a field attribute. We still have to keep shims around for old deprecated fields to avoid breaking migrations for example.

> Anyhow, here is my suggestion. I will remove the returning kwarg from the constructor. That way you can't set it at all. It will remain something only exposed to custom fields. Sounds good?

If we can avoid using a kwarg for this I think it's a safe way forward. I think it'd even be better if all the returning logic would be pushed down to the compiler without defining any model field attributes for now. Having the compiler automatically build it's `RETURNING` clause based on expressions that are db generated would be even better IMO.

codingjoe replied:
> @codingjoe my main concern with adding a model field kwargs is the probability we'll have to deprecate it in the future in favor of something that is more though of and in line with the previous discussions (django/deps#36, #5904, #2149). For example the ORM usually tries to abstract implementation details and prefix database related options with db_*.

Good point, I will rename the attributes to `db_returning`. That does seem be a more descriptive name.

> If we can avoid using a kwarg for this I think it's a safe way forward. I think it'd even be better if all the returning logic would be pushed down to the compiler without defining any model field attributes for now. Having the compiler automatically build it's RETURNING clause based on expressions that are db generated would be even better IMO.

I believe the change I made actually goes into that direction, even if not fully meeting that requirement. See, previously the compiler did know what a model has to have primary key and made sure to return it. Now the compiler is just being told, "hey, insert that! oh and while you are at it, I what is the value of the `id` field?"
The `RETURNING` statement is compiled by the compiler, the only thing that the compiler does not know, is which fields should be returned and which not. I know if would be super elegant to go with the `.as_sql` approach for default values. As cool as that might be, I believe it should be done in a separate commit and preferably after gaining some experience with some actual implementations of this feature.

charettes commented on django/db/models/sql/compiler.py:
```diff
@@ -1275,27 +1275,25 @@ def as_sql(self):
         # necessary and it should be possible to use placeholders and
         # expressions in bulk inserts too.
         can_bulk = (not self.return_id and self.connection.features.has_bulk_insert)
-
```
Unnecessary change.

timgraham commented:
Here are the failures on Oracle:
```
======================================================================
ERROR: test_json_serializer (serializers.test_data.SerializerDataTests)
partial(func, *args, **keywords) - new function with partial application
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/serializers/test_data.py", line 409, in serializerTest
    serialized_data = serializers.serialize(format, objects, indent=2)
  File "/media/evo/djangō/django/core/serializers/__init__.py", line 128, in serialize
    s.serialize(queryset, **options)
  File "/media/evo/djangō/django/core/serializers/base.py", line 115, in serialize
    self.end_object(obj)
  File "/media/evo/djangō/django/core/serializers/json.py", line 53, in end_object
    json.dump(self.get_dump_object(obj), self.stream, **self.json_kwargs)
  File "/media/evo/djangō/django/core/serializers/python.py", line 38, in get_dump_object
    data["pk"] = self._value_from_field(obj, obj._meta.pk)
  File "/media/evo/djangō/django/core/serializers/python.py", line 47, in _value_from_field
    return value if is_protected_type(value) else field.value_to_string(obj)
  File "/media/evo/djangō/django/db/models/fields/__init__.py", line 1281, in value_to_string
    return '' if val is None else val.isoformat()
AttributeError: 'str' object has no attribute 'isoformat'

======================================================================
ERROR: test_python_serializer (serializers.test_data.SerializerDataTests)
partial(func, *args, **keywords) - new function with partial application
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/serializers/test_data.py", line 409, in serializerTest
    serialized_data = serializers.serialize(format, objects, indent=2)
  File "/media/evo/djangō/django/core/serializers/__init__.py", line 128, in serialize
    s.serialize(queryset, **options)
  File "/media/evo/djangō/django/core/serializers/base.py", line 115, in serialize
    self.end_object(obj)
  File "/media/evo/djangō/django/core/serializers/python.py", line 32, in end_object
    self.objects.append(self.get_dump_object(obj))
  File "/media/evo/djangō/django/core/serializers/python.py", line 38, in get_dump_object
    data["pk"] = self._value_from_field(obj, obj._meta.pk)
  File "/media/evo/djangō/django/core/serializers/python.py", line 47, in _value_from_field
    return value if is_protected_type(value) else field.value_to_string(obj)
  File "/media/evo/djangō/django/db/models/fields/__init__.py", line 1281, in value_to_string
    return '' if val is None else val.isoformat()
AttributeError: 'str' object has no attribute 'isoformat'

======================================================================
ERROR: test_xml_serializer (serializers.test_data.SerializerDataTests)
partial(func, *args, **keywords) - new function with partial application
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/serializers/test_data.py", line 411, in serializerTest
    for obj in serializers.deserialize(format, serialized_data):
  File "/media/evo/djangō/django/core/serializers/xml_serializer.py", line 168, in __next__
    return self._handle_object(node)
  File "/media/evo/djangō/django/core/serializers/xml_serializer.py", line 181, in _handle_object
    node.getAttribute('pk'))
  File "/media/evo/djangō/django/db/models/fields/__init__.py", line 1246, in to_python
    params={'value': value},
django.core.exceptions.ValidationError: ["'2006-06-16 00:00:00' value has an invalid date format. It must be in YYYY-MM-DD format."]

======================================================================
ERROR: test_yaml_serializer (serializers.test_data.SerializerDataTests)
partial(func, *args, **keywords) - new function with partial application
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/serializers/test_data.py", line 409, in serializerTest
    serialized_data = serializers.serialize(format, objects, indent=2)
  File "/media/evo/djangō/django/core/serializers/__init__.py", line 128, in serialize
    s.serialize(queryset, **options)
  File "/media/evo/djangō/django/core/serializers/base.py", line 115, in serialize
    self.end_object(obj)
  File "/media/evo/djangō/django/core/serializers/python.py", line 32, in end_object
    self.objects.append(self.get_dump_object(obj))
  File "/media/evo/djangō/django/core/serializers/python.py", line 38, in get_dump_object
    data["pk"] = self._value_from_field(obj, obj._meta.pk)
  File "/media/evo/djangō/django/core/serializers/python.py", line 47, in _value_from_field
    return value if is_protected_type(value) else field.value_to_string(obj)
  File "/media/evo/djangō/django/db/models/fields/__init__.py", line 1281, in value_to_string
    return '' if val is None else val.isoformat()
AttributeError: 'str' object has no attribute 'isoformat'

======================================================================
FAIL: test_inline_formsets_with_custom_pk (model_formsets.tests.ModelFormsetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/model_formsets/tests.py", line 787, in test_inline_formsets_with_custom_pk
    self.assertEqual(book1.pk, 77777)
AssertionError: '77777' != 77777

======================================================================
FAIL: test_two_level_foreign_keys (model_fields.test_uuid.TestAsPrimaryKey)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/model_fields/test_uuid.py", line 193, in test_two_level_foreign_keys
    self.assertIsInstance(gc.uuidchild_ptr_id, uuid.UUID)
AssertionError: '05dcf43b439849acaa4d295384cea021' is not an instance of <class 'uuid.UUID'>

======================================================================
FAIL: test_hidden_no_data_found_exception (backends.oracle.tests.TransactionalTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/evo/djangō/tests/backends/oracle/tests.py", line 82, in test_hidden_no_data_found_exception
    Square.objects.create(root=2, square=4)
  File "/opt/python3.7.0/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/media/evo/djangō/django/test/testcases.py", line 675, in _assert_raises_or_warns_cm
    yield cm
AssertionError: DatabaseError not raised
```
@felixxm might have some insight about the implementation there.

felixxm commented:
I will check Oracle failures at the weekend.

codingjoe commented:
> @codingjoe Thanks for this patch 👍 Sorry, but I'm a bit confused 😕 My understanding is that this change should allow for returning multiple fields (via `RETURNING` clause) during `INSERT`, but we're going to restrict this only to PKs (at least for now). Am I right?

No, not exactly. It will not be restricted to PKs on Postgres. Of all other DBs it will, since they don't support multiple return values. (Or in the case of Oracle, the db wrapper does not support it.)

I'm asking because in [`return_insert_columns()`](https://github.com/django/django/pull/9983/files#diff-ea23a0501b6f65fb447da98953a35dafR318) we have `SQL` with multiple columns and in the same time `cursor._insert_id_var` is always a single parameter and [fetch_returned_insert_columns()](https://github.com/django/django/pull/9983/files#diff-ea23a0501b6f65fb447da98953a35dafR229) returns a list with a single value 🤔 IMO we should change `_insert_id_var` to the list of variables `_insert_vars` and fetch all of them in the `fetch_returned_insert_columns()` also we should probably run db converters on results (see [Variable.outconverter](https://cx-oracle.readthedocs.io/en/latest/variable.html#Variable.outconverter)) and make some changes in the `InsertReturnParameter.types` (see [DatabaseIntrospection.data_types_reverse](https://github.com/django/django/blob/master/django/db/backends/oracle/introspection.py#L15-L28)). If you want I can prepare some edits.

I think I understood what you mean. It still is not consistent, mainly because some dbs don't support more than once return value. I would be interested to see what solution you have in mind. It might also be possible to just have multiple different `InsertCompiler`s for each db and not one, that tries to do everything, as it currently is.

codingjoe commented:
I am going to split this PR once more. I will pull out the Oracle part into a separate PR.

codingjoe commented:
I ran the Oracle tests locally, they all passed.

felixxm commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm @timgraham are we good here? With 2.2 
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@timgraham @felixxm I just moved the release lots to 3.0 and since 2.2 is out, maybe we can get this on the road?

apollo13 commented:
@codingjoe I think this pull request misses a test that tests that the actual returning of the values works as expected (for create and bulk_create for instance)

codingjoe commented:
@apollo13 yes, @charettes was asking for a couple more tests too. I will try to find some time this weekend and go nuts on tests.

carltongibson commented on docs/releases/3.0.txt:
```diff
@@ -185,6 +185,9 @@ Models
   :class:`~django.db.models.functions.Trunc` database functions determines the
   treatment of nonexistent and ambiguous datetimes.
 
+* Added support for non-integer primary keys to Oracle. Vales are returned by
```
`Values`

codingjoe replied:
resolved in #11462

codingjoe commented:
> Just wondering: is `Added support for non-integer primary keys to Oracle.` strictly tied to this PR?

@carltongibson no it is not. I will move it into a separate PR. This will make reviewing this much simpler and allow us to evaluate the corret test coverage.

carltongibson commented:
Super. Thanks @codingjoe! 👍

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

felixxm commented:
Rebase is required after #11462 :pray:

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented on django/db/backends/oracle/utils.py:
```diff
@@ -10,16 +10,22 @@ class InsertVar:
     def __init__(self, field):
         internal_type = getattr(field, 'target_field', field).get_internal_type()
-        self.db_type = self.types.get(internal_type, int)
+        self.db_type = self.types.get(internal_type, str)
```
@felixxm with the `InsertVar` actually now being used in more places, I need to make a bunch of adjustment to make this work for all scenarios.
I decided to add all integer types explicitly and use `str` as a fallback. This made more senes, since the majority of fields like `UUID` or `GenericIPAddress` work just fine with strings. I believe this will also be better suited for any third party library, that might implement it's own fields.

The test that really test all different cases here is: `serializers.test_data.SerializerDataTests`. It inserts all different field types as primary keys.

felixxm replied:
Agreed, can we remove `CharField` from `types` and move this adjustment to a separate commit, e.g. `Refs #29444 -- Made db.backends.oracle.utils.InsertVar use str as default.`

codingjoe commented:
@felixxm I left you one comment about necessary adjustments that I made. Besides that I hope this will finally ready for check in. It would be so cool to get to possibility to implement database defaults with Django 3.

felixxm commented on django/db/backends/oracle/utils.py:
```diff
@@ -10,16 +10,22 @@ class InsertVar:
         'CharField': str,
         'DateTimeField': Database.TIMESTAMP,
-        'DateField': Database.DATETIME,
+        'DateField': Database.Date,
```
Please revert this change.

codingjoe replied:
@felixxm the tests fail if I do.

felixxm replied:
That's weird :thinking: Let me check.

felixxm replied:
OK, it's my mistake from #11462 :disappointed: Please keep this change in the main commit and use `Database.DATE`.

codingjoe replied:
I put this into a separate commit as well.

codingjoe commented:
buildbot, test on oracle.

felixxm commented on tests/queries/test_db_returning.py:
```diff
@@ -0,0 +1,43 @@
+        for obj in objs:
+            with self.subTest(obj=obj):
+                self.assertTrue(obj.pk)
+                self.assertIsInstance(obj.created, datetime.datetime)
```
@codingjoe I'm going to check why `test_insert` and `test_bulk_insert` fail on Oracle with the current implementation :thinking: .

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm I applied the change you suggested. I was considering introducing this in the next part of my feature journey, but this is probably the better place.

felixxm commented:
@codingjoe Sorry for misleading you :confused:, but I was trying to say that we should reduce the scope of this PR. Current `db_returning` implementation impacts all primary keys without defaults IMO this **should not** happen for most of them (except auto fields) because it can cause performance issue and it's not necessary as long as we don't support database defaults. So, my proposition is to return `False` by default and `True` for `AutoField`'s:

```python
class Field(RegisterLookupMixin):
    ...
    @property
    def db_returning(self):
        return False
    ...

class AutoField(Field):
    ...
    @property
    def db_returning(self):
        return True
    ...
```

`db_returning` that would check type of `default` was a proposition for a PR with adding official support for database defaults.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm haha, ok, I see what you mean. I made the changes.

pope1ni commented on django/db/backends/oracle/operations.py:
```diff
@@ -334,8 +336,20 @@ def regex_lookup(self, lookup_type):
+        params = ()
+        into_params = []
+        for field in fields:
+            columns.append("%s.%s" % (self.quote_name(field.model._meta.db_table), self.quote_name(field.column)))
```
Please use single quotes everywhere in this method.

pope1ni commented on django/db/backends/oracle/utils.py:
```diff
@@ -10,16 +10,21 @@ class InsertVar:
     """
     types = {
+        'AutoField': int,
+        'BigAutoField': int,
```
Please also add the new `SmallIntegerField` to this.

codingjoe replied:
and `SmallAutoField`

pope1ni replied:
I must have been half asleep when reviewing - sorry! I thought I had said `SmallAutoField`, but I was still right in suggesting the other 😅

pope1ni commented on django/db/backends/oracle/operations.py:
```diff
@@ -334,8 +336,20 @@ def regex_lookup(self, lookup_type):
+            into_params.append("%s")
+        return "RETURNING %s INTO %s" % (
+            ", ".join(columns),
+            ", ".join(into_params)
```
This could be `', '.join('%s' for i in len(columns)),` to avoid needing to create `into_params` up front.

codingjoe replied:
`TypeError: 'int' object is not iterable` ;) That could be fixed, but it is very consistent avoids calculating the list size.

pope1ni replied:
Sorry, I meant `range()`, not `len()`. (Or just ditch the `len()` and iterate over the elements.)

pope1ni commented on django/db/backends/oracle/operations.py:
```diff
@@ -334,8 +336,20 @@ def regex_lookup(self, lookup_type):
+        for field in fields:
+            columns.append("%s.%s" % (self.quote_name(field.model._meta.db_table), self.quote_name(field.column)))
+            params += (InsertVar(field),)
+            into_params.append("%s")
```
Simplify to the following?
```python
columns = ['%s.%s' % (self.quote_name(f.model._meta.db_table), self.quote_name(f.column)) for f in fields]
params = tuple(InsertVar(f) for f in fields)
```

codingjoe replied:
I guess simple is a bit subjective, yet I agree with you on this one. But, it's actually slower, since you need to loop over fields twice. Therefore, I'd like to keep this as is.

pope1ni replied:
I guess I find comprehensions more legible than loops like this, especially with single item tuple appends which are syntactically ugly. Comprehensions are fast with dedicated operations in bytecode, calling methods like `.append()` multiple times in the loop is slow.

In terms of performance, it should be negligible anyway as I expect this to be called once before the query executes, not per row returned, and typically I expect the number of fields to be very small...

pope1ni commented on django/db/backends/postgresql/operations.py:
```diff
@@ -236,8 +235,17 @@ def last_executed_query(self, cursor, sql, params):
+                self.quote_name(field.model._meta.db_table),
+                self.quote_name(field.column)
+            )
+            columns.append(db_column)
```
Please make this a list comprehension.

pope1ni commented on django/db/backends/postgresql/operations.py:
```diff
@@ -236,8 +235,17 @@ def last_executed_query(self, cursor, sql, params):
+                self.quote_name(field.column)
+            )
+            columns.append(db_column)
+        return "RETURNING %s" % ", ".join(columns), ()
```
Single quotes please.

pope1ni commented on django/db/models/fields/__init__.py:
```diff
@@ -953,6 +959,9 @@ def rel_db_type(self, connection):
     def validate(self, value, model_instance):
         pass
 
+    def db_returning(self):
```
Should be decorated with `@property`?

pope1ni commented on django/db/models/sql/compiler.py:
```diff
@@ -1331,24 +1330,34 @@ def as_sql(self):
-            )
+                return self.connection.ops.fetch_returned_insert_columns(cursor)
+            meta = self.query.get_meta()
+            if meta.pk.get_internal_type() in ['AutoField', 'BigAutoField', 'SmallAutoField']:
```
Prefer a set (or tuple) to list for containment check.

pope1ni replied:
This feels like a kludge. Is it dependant on something like the changes in #3776?

codingjoe replied:
Good catch, a set is particularly fast if the element is not contained, which is the larger case.

And yes, it's a kludge, but no one has come up with a better solution yet. It would need some major refactoring of the database backends. Which I am happy to do, but not as part of this PR.

pope1ni commented on docs/releases/3.0.txt:
```diff
@@ -392,8 +392,14 @@ backends.
-  ``field`` argument with the model field.
+* ``DatabaseOperations.fetch_returned_insert_id()`` is renamed to
+  ``fetch_returned_insert_columns()`` and returns a list instead of single
+  value.
```
```
… is replaced by … which returns a list of values generated by the query during execution instead of a single value, which was typically the auto-incremented primary key.
```

pope1ni commented on docs/releases/3.0.txt:
```diff
@@ -392,8 +392,14 @@ backends.
+  ``fetch_returned_insert_columns()`` and returns a list instead of single
+  value.
+
+* ``DatabaseOperations.return_insert_id()`` is renamed to
```
```… is replaced by …```

pope1ni commented on docs/releases/3.0.txt:
```diff
@@ -392,8 +392,14 @@ backends.
+
+* ``DatabaseOperations.return_insert_id()`` is renamed to
+  ``DatabaseOperations.return_insert_columns()`` and expects a new ``fields``
+  argument, which is a iterable of fields to be returned after insert. Usually
```
`a iterable` → `an iterable`

pope1ni commented on tests/queries/models.py:
```diff
@@ -730,3 +731,16 @@ class Meta:
+        kwargs.setdefault('default', Now)
+        super().__init__(*args, **kwargs)
+
+    def db_returning(self):
```
Should be decorated with `@property`?

codingjoe commented:
buildbot, test on oracle.

auvipy commented on django/db/models/fields/__init__.py:
```diff
@@ -718,6 +718,12 @@ def unique(self):
     def db_tablespace(self):
         return self._db_tablespace or settings.DEFAULT_INDEX_TABLESPACE
 
```
seems there are conflicts here

codingjoe replied:
I'll rebase everything.

felixxm commented:
@codingjoe I've started polishing this patch. I pushed minor edits to docs and tests simplification. I also found few unresolved comments, first of all I wanted to address [Simon's comment](https://github.com/django/django/pull/9983#discussion_r243740547). Can you take a look at my proposition 13477049aa6b959a98d6fb941a08edfa13046b31? If you're fine with it, I will squash commits and move forward to remaining comments.

codingjoe commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm thanks for the great work. It looks correct, but I need to run Oracle first.  I see you are having trouble triggering the build too. I am currently running it locally... but its sloooooow ;)

In any event, I don't mind you tying things up here, in fact I really appreciate it.

felixxm commented:
Traging an Oracle build is not so important at the moment because `test_db_returning` are failing on Oracle (IMO they should pass without a vendor check). I will try to fix them.

codingjoe commented:
> Traging an Oracle build is not so important at the moment because test_db_returning are failing on Oracle (IMO they should pass without a vendor check). I will try to fix them.

@felixxm don't. It does not and is not supposed to work on oracle. Tho we can build statements with multiple returning columns, the Oracle db driver does actually not support fetching multiple returning columns. This is a PG ONLY feature unless Oracle gets it together and implements that feature in their python driver.

felixxm commented:
@codingjoe I'm a bit confused :confused:. Can you provide any source about this restriction in `cx_Oracle`?

codingjoe commented:
@felixxm I might just be confusing that. It's been a while and I don't use Oracle except for this patch.
I believe I might be confusing this, with the recently added batch support, see also https://blogs.oracle.com/opal/python-cx_oracle-63-supports-dml-returning-for-batch-statement-execution

That being said, the current implementation does sets the parameter on the cursor https://github.com/django/django/blob/003bb34b218adb23d1a7e67932a6ba9b3c4dcc81/django/db/backends/oracle/utils.py#L33
and accesses it later over here https://github.com/django/django/blob/4137fc2efce2dde48340728b8006fc6d66b9e3a5/django/db/backends/oracle/operations.py#L252

This does limit the current implementation of Oracle to a single column. Not saying that couldn't be done, but I am not sure I am the one doing it, since I really just want to implement this for Postgres.

felixxm commented:
@codingjoe Thanks for links, now it's clear to me. I'm going to revert a part of Oracle changes and leave a ticket open. Let's work on them for Django 3.1, I don't have enough time to polish them before Monday. I will try to add support for Oracle (with `cx_Oracle` 6.3+) before Django 3.1 feature freeze :crossed_fingers:

felixxm commented:
I added DB feature, raised an error on Oracle, simplified Oracle's changes and used DB feature in tests. Hopefully I will finish this tomorrow :crossed_fingers:

codingjoe commented:
> I added DB feature, raised an error on Oracle, simplified Oracle's changes and used DB feature in tests. Hopefully I will finish this tomorrow 🤞

Thank you for your @felixxm effort. I understand your concern, but to be honest, at this point I am a bit tired. I have been working on this for 3 years now and the scope keeps expanding. So if we can just gracefully fail on the Oracle end (for now), I'd be over the moon if this still makes it into the next release.

felixxm commented:
I pushed edits. IMO it's ready.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.

pope1ni commented on django/db/models/query.py:
```diff
@@ -1203,21 +1204,22 @@ def _batched_insert(self, objs, fields, batch_size, ignore_conflicts=False):
                 )
-                if isinstance(inserted_id, list):
-                    inserted_ids.extend(inserted_id)
+                if isinstance(inserted_columns, list):
```
Can we always assume we are getting a list returned? I believe `SQLInsertCompiler.execute_sql()` always returns a list now? (Although we may need a note in the backward incompatibility section for database backends?)

If all that were the case, we could possibly condense this whole section to something like the following:
```python
returning_fields = None
bulk_return = connections[self.db].features.can_return_rows_from_bulk_insert
if bulk_return and not ignore_conflicts:
    returning_fields = self.model._meta.db_returning_fields
return list(itertools.chain.from_iterable(
    self._insert(
        item, fields=fields, using=self.db,
        returning_fields=returning_fields,
        ignore_conflicts=ignore_conflicts,
    )
    for item in (objs[i:i + batch_size] for i in range(0, len(objs), batch_size)):
))
```

codingjoe replied:
Sadly that does not work, since execute does always return a list, but it can be a list of tuples. Or a list of a list of a tuple.

codingjoe commented on django/db/models/query.py:
```diff
@@ -485,8 +485,9 @@ def bulk_create(self, objs, batch_size=None, ignore_conflicts=False):
-                    obj_without_pk.pk = pk
+                for obj_without_pk, results in zip(objs_without_pk, ids):
+                    for result, field in zip(results, self.model._meta.db_returning_fields):
+                        setattr(obj_without_pk, field.attname, result)
```
I added a commit that should resolve both issues.

felixxm replied:
`RETURNING` from `UPDATE` is out of this ticket scope.

felixxm commented:
(grumbling) I'm not sure if we're going into the right the direction I wanted to fix remaining Oracle failure and merge this but now at least extra tests are required for new scenarios. I'm not sure if I will find few more hours to finish this before a feature freeze :disappointed:

codingjoe commented:
buildbot, test on oracle.

felixxm commented:
It will not work on Oracle, I didn't push my fix.

felixxm commented:
I will try to do this later...

codingjoe commented:
@felixxm maybe I can help, what did you have in mind?

auvipy commented:
what about pushing feature freeze 1 week further?

felixxm commented:
@auvipy We will not postpone a feature freeze, it doesn't solve anything. Everyone would like to include their features, schedule is schedule.

codingjoe commented:
@felixxm I am implementing `can_return_multiple_columns_from_insert` support for oracle as we speak. It's almost working, I am just writing tests. 🤞

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
buildbot, test on oracle.

codingjoe commented:
@felixxm I can't seem to trigger buildbot to build oracle builds, but locally the tests seem to pass. So yes, this feature is now also available to Oracle. You might want to update the release not since I updated `fetch_returned_insert_columns` signature. I would leave that to you tho, since I am dyslexic and I want to cause any more trouble ;)

codingjoe commented:
buildbot, test on oracle.

felixxm commented:
@codingjoe Thanks, I'm not sure if that was a good decision because I may not find enough time to check new code, before Monday :| We'll see.

felixxm commented:
... but maybe I will just move the last commit to a new PR.

codingjoe commented:
@felixxm whatever you deem best.

pope1ni commented:
Sorry, sorry! :scream: 

I didn't mean to be awkward. I think if we keep this small for 3.0, we can build on it later. The only thing for now would be to consider the `objs_with_pk` branch in `.bulk_create()` so that it behaves consistently. As long as the returning of id works as before this should be fine.

For 3.1 we can then look into the other improvements that would include `UPDATE ... RETURNING ...` and support for Oracle... And hopefully then some of the other cool things we can build on top of that... :guitar:

felixxm commented:
@pope1ni don't be, I'm just grumpy 😁 and yes that's my plan for Monday, i.e. fix `bulk_create()`, add test for this scenario, fix one failure on Oracle (this is already ready on my local branch), and move support for Oracle to a separate PR.

codingjoe commented:
You lost me, what's not working at this point? Are we talking about fetching IDs from bulk inserts in Oracle? Which is a separate feature, right? So this remains ready for check in (minus whatever you want to (re)move to a separate PR)?

charettes commented on django/db/models/fields/__init__.py:
```diff
@@ -718,6 +718,14 @@ def unique(self):
     def db_tablespace(self):
         return self._db_tablespace or settings.DEFAULT_INDEX_TABLESPACE
 
+    @property
```
Any reason to make this a property instead of a class level attribute?

felixxm replied:
The default implementation will be more complicated in the future when we will add db defaults. I think we can leave it is as a property.

charettes commented on django/db/models/fields/__init__.py:
```diff
@@ -2326,6 +2334,10 @@ def deconstruct(self):
     def validate(self, value, model_instance):
         pass
 
+    @property
```
Ditto?

charettes commented on tests/queries/models.py:
```diff
@@ -730,3 +731,17 @@ class Meta:
+        super().__init__(*args, **kwargs)
+
+    @property
+    def db_returning(self):
```
Again, this could be a class level attribute.

felixxm replied:
Agreed.

felixxm commented:
@codingjoe I moved failing Oracle test from `backends/oracle/tests.py` to `queries/test_db_returning.py`, added test for `bulk_create()` with and without ID and removed solution for `cx_Oracle 6.3+`.

I think we should discuss solution for Oracle in a separate PR. First of all Django supports also `cx_Oracle` < 6.3 so it will not work for all supported drivers, moreover I don't have enough time today to review, test and polish this. Thanks for understanding. If you don't have time or energy to submit a separate PR I can do this for you.

Thanks again for your hard work and patience :rocket:

felixxm commented:
buildbot, test on oracle.

felixxm commented:
buildbot, test on oracle.
